iceoryx_hoofs/source/posix_wrapper/shared_memory_object/shared_memory.cpp

```diff
@@ -33,73 +33,151 @@ namespace iox
 {
 namespace posix
 {
-// NOLINTNEXTLINE(readability-function-size) todo(iox-#832): make a struct out of arguments
-SharedMemory::SharedMemory(const Name_t& name,
-                           const AccessMode accessMode,
-                           const OpenMode openMode,
-                           const mode_t permissions,
-                           const uint64_t size) noexcept
+static int getOflagsFor(const AccessMode accessMode, const OpenMode openMode) noexcept
 {
-    m_isInitialized = true;
+    int oflags = 0;
+    oflags |= (accessMode == AccessMode::READ_ONLY) ? O_RDONLY : O_RDWR;
+    oflags |= (openMode != OpenMode::OPEN_EXISTING) ? O_CREAT : 0;
+    oflags |= (openMode == OpenMode::EXCLUSIVE_CREATE) ? O_EXCL : 0;
+    return oflags;
+}
+
+cxx::string<SharedMemory::Name_t::capacity() + 1> addLeadingSlash(const SharedMemory::Name_t& name) noexcept
+{
+    cxx::string<SharedMemory::Name_t::capacity() + 1> nameWithLeadingSlash = "/";
+    nameWithLeadingSlash.append(cxx::TruncateToCapacity, name);
+    return nameWithLeadingSlash;
+}
+
+cxx::expected<SharedMemory, SharedMemoryError> SharedMemoryBuilder::create() noexcept
+{
+    auto printError = [this] {
+        std::cerr << "Unable to create shared memory with the following properties [ name = " << m_name
+                  << ", access mode = " << ACCESS_MODE_STRING[static_cast<uint64_t>(m_accessMode)]
+                  << ", open mode = " << OPEN_MODE_STRING[static_cast<uint64_t>(m_openMode)]
+                  << ", mode = " << std::bitset<sizeof(mode_t)>(static_cast<mode_t>(m_filePermissions))
+                  << ", sizeInBytes = " << m_size << " ]" << std::endl;
+    };
+
+
     // on qnx the current working directory will be added to the /dev/shmem path if the leading slash is missing
-    if (name.empty())
+    if (m_name.empty())
     {
         std::cerr << "No shared memory name specified!" << std::endl;
-        m_isInitialized = false;
-        m_errorValue = SharedMemoryError::EMPTY_NAME;
+        return cxx::error<SharedMemoryError>(SharedMemoryError::EMPTY_NAME);
     }
-    // empty case handled above, so it's fine to get first element
-    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
-    else if (name.c_str()[0] != '/')
+
+    if (!cxx::isValidFileName(m_name))
     {
-        std::cerr << "Shared memory name must start with a leading slash!" << std::endl;
-        m_isInitialized = false;
-        m_errorValue = SharedMemoryError::NAME_WITHOUT_LEADING_SLASH;
+        std::cerr << "Shared memory requires a valid file name (not path) as name and \"" << m_name
+                  << "\" is not a valid file name" << std::endl;
+        return cxx::error<SharedMemoryError>(SharedMemoryError::INVALID_FILE_NAME);
     }
 
-    if (m_isInitialized)
+    auto nameWithLeadingSlash = addLeadingSlash(m_name);
+
+    bool hasOwnership = (m_openMode == OpenMode::EXCLUSIVE_CREATE || m_openMode == OpenMode::PURGE_AND_CREATE
```

Comment:
Could you move this down to line 128, shortly before it is actually used?

Reply:
No problemo.